libr/bin/dwarf.c

```diff
@@ -734,69 +800,55 @@ R_API int r_bin_dwarf_parse_line_raw2(const RBin *a, const ut8 *obuf,
 	return true;
 }
 
-#define READ_BUF(x,y) if (idx+sizeof(y)>=len) { return false;} \
-	(x)=*(y*)buf; idx+=sizeof(y);buf+=sizeof(y)
-
 R_API int r_bin_dwarf_parse_aranges_raw(const ut8 *obuf, int len, FILE *f) {
 	ut32 length, offset;
 	ut16 version;
-	ut32 debug_info_offset;
-	ut8 address_size, segment_size;
+	ut32 cu_offset;
+	ut8 addr_size, seg_size;
 	const ut8 *buf = obuf;
-	int idx = 0;
 
 	if (!buf || len< 4) {
 		return false;
 	}
 
-	READ_BUF (length, ut32);
-	if (f) {
-		printf("parse_aranges\n");
-		printf("length 0x%x\n", length);
-	}
-
-	if (idx+12>=len)
-		return false;
-
-	READ_BUF (version, ut16);
-	if (f) {
-		printf("Version %d\n", version);
-	}
-
-	READ_BUF (debug_info_offset, ut32);
-	if (f) {
-		fprintf (f, "Debug info offset %d\n", debug_info_offset);
-	}
-
-	READ_BUF (address_size, ut8);
-	if (f) {
-		fprintf (f, "address size %d\n", (int)address_size);
-	}
+	// FIXME endian
+	data_xtor ex;
+	data_xtor_init (&ex, obuf, len, false);
+	st64 off = 0;
 
-	READ_BUF (segment_size, ut8);
+	length = data_xtor_u32 (&ex, &off);
+	version = data_xtor_u16 (&ex, &off);
+	cu_offset = data_xtor_u32 (&ex, &off);
+	addr_size = data_xtor_u8 (&ex, &off);
+	seg_size = data_xtor_u8 (&ex, &off);
 	if (f) {
-		fprintf (f, "segment size %d\n", (int)segment_size);
+		fprintf (f, "parse_aranges\n");
+		fprintf (f, "length 0x%x\n", length);
+		fprintf (f, "Version %d\n", version);
+		fprintf (f, "Debug info offset %u\n", cu_offset);
+		fprintf (f, "address size %d\n", (int)addr_size);
+		fprintf (f, "segment size %d\n", (int)seg_size);
 	}
 
-	offset = segment_size + address_size * 2;
+	offset = seg_size + addr_size * 2;
 
 	if (offset) {
 		ut64 n = (((ut64) (size_t)buf / offset) + 1) * offset - ((ut64)(size_t)buf);
-		if (idx+n>=len) {
+		if (off+n>=len) {
 			return false;
 		}
 		buf += n;
-		idx += n;
+		off += n;
 	}
 
-	while ((buf - obuf) < len) {
-		ut64 adr, length;
-		if ((idx+8)>=len) {
+	while (off < len) {
+		st64 old = off;
+		ut64 address = data_xtor_unsigned (&ex, &off, addr_size);
+		ut64 length = data_xtor_unsigned (&ex, &off, addr_size);
+		if (off - old != addr_size * 2) {
 			break;
 		}
-		READ_BUF (adr, ut64);
-		READ_BUF (length, ut64);
-		if (f) printf("length 0x%"PFMT64x" address 0x%"PFMT64x"\n", length, adr);
+		if (f) printf("length 0x%"PFMT64x" address 0x%"PFMT64x"\n", length, address);
```

Comment:
add braces for this if and add space before (

Reply:
also, dont use printf. if this is a debug message use r_cons_printf, but other functions in this file seems to use `		fprintf (f, "segment size %d\n", (int)seg_size);`